BoogieLang/Ast_Cfg_Transformation.thy

```diff
@@ -0,0 +1,2280 @@
+  shows "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>expr, ns1\<rangle> \<Down> BoolV (True)"  
+  using assms
+  proof cases
+    case (RedUnOp v)
+    from this obtain b1 where "v = LitV (LBool b1)"
+      by (metis (no_types, lifting) map_option_eq_Some option.simps(3) unop_eval.simps(1) unop_eval_val.elims unop_not.elims)   
+    from this RedUnOp have 
+      expand1: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>expr,ns1\<rangle> \<Down> (LitV (LBool b1))" and 
+      expand2: "unop_eval_val unop.Not (LitV (LBool b1)) = Some (BoolV False)"
+      by auto
+    then show ?thesis by fastforce
+  qed
+
+lemma true_equals_not_false:
+  assumes "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>expr, ns1\<rangle> \<Down> BoolV True"
+  shows "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>UnOp unop.Not expr, ns1\<rangle> \<Down> BoolV (False)"
+  using assms by (simp add: red_expr_red_exprs.intros(5))
+
+lemma false_equals_not_true:
+  assumes "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>expr, ns1\<rangle> \<Down> BoolV False"
+  shows "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>UnOp unop.Not expr, ns1\<rangle> \<Down> BoolV (True)"
+  using assms by (simp add: red_expr_red_exprs.intros(5))
+
+lemma equiv_preserves_value:
+  assumes "a \<sim> b"
+  and "red_expr A \<Lambda> \<Gamma> \<Omega> a ns (BoolV boolean)"
+  shows "red_expr A \<Lambda> \<Gamma> \<Omega> b ns (BoolV boolean)"
+  using assms
+proof cases
+  case (neg_refl e1)
+  then show ?thesis using assms by simp
+next
+  case neg_equiv1
+  then show ?thesis using assms by (metis (full_types) RedLit not_true_equals_false val_elim)
+next
+  case neg_equiv2
+  then show ?thesis using assms by (metis (full_types) RedLit not_false_equals_true val_elim)
+next
+  case double_neg
+  then show ?thesis using assms by (metis (full_types) not_false_equals_true not_true_equals_false)
+next
+  case (neg_eq e1 e2)
+  from this assms have unop_red: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>UnOp unop.Not (e1 \<guillemotleft>Eq\<guillemotright> e2), ns\<rangle> \<Down> BoolV boolean" by simp
+  show ?thesis
+  proof (cases boolean)
+    case True
+    from this assms neg_eq have 
+      "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>UnOp unop.Not (e1 \<guillemotleft>Eq\<guillemotright> e2), ns\<rangle> \<Down> BoolV True" by simp
+    hence eq_false: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>(e1 \<guillemotleft>Eq\<guillemotright> e2), ns\<rangle> \<Down> BoolV False" by (metis not_true_equals_false)
+    from this obtain v1 v2 where
+      redE1: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>e1, ns\<rangle> \<Down> v1" and
+      redE2: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>e2, ns\<rangle> \<Down> v2" by blast
+    from this eq_false have "v1 \<noteq> v2" 
+      using RedBinOp_case
+      by (smt (verit, del_insts) RedBinOp_case binop_eval_val.simps(1) expr_eval_determ(1) lit.inject(1) option.inject val.inject(1))
```

Comment:
It would be good to get rid of smt proofs if possible (same comment for other smt proofs). They are not very stable (for example, on my Windows machine sometimes smt proofs go through and sometimes they don't).